aea/configurations/components.py

```diff
@@ -0,0 +1,178 @@
+    ComponentConfiguration,
+    ComponentId,
+    ComponentType,
+    ConnectionConfig,
+    PublicId,
+)
+from aea.helpers.base import _SysModules, load_init_modules, load_module
+
+logger = logging.getLogger(__name__)
+
+
+def component_class_from_type(component_type: ComponentType) -> Type["Component"]:
+    """Get component class from component type."""
+    from aea.protocols.base import Protocol
+    from aea.connections.base import Connection
+    from aea.skills.base import Skill
+
+    if component_type == ComponentType.PROTOCOL:
+        return Protocol
+    elif component_type == ComponentType.CONNECTION:
+        return Connection
+    elif component_type == ComponentType.SKILL:
+        return Skill
+    elif component_type == ComponentType.CONTRACT:
+        # TODO
+        raise NotImplementedError
+    else:
+        raise ValueError
+
+
+class Component(ABC):
+    """Abstract class for an agent component."""
+
+    def __init__(self, configuration: ComponentConfiguration, is_vendor: bool = False):
+        """
+        Initialize a package.
+
+        :param configuration: the package configuration.
+        :param is_vendor: whether the package is vendorized.
+        """
+        self._configuration = configuration
+        self._directory = None  # type: Optional[Path]
+        self._is_vendor = is_vendor
+
+        # mapping from import path to module object
+        # the keys are dotted paths of Python modules.
+        self.importpath_to_module = {}  # type: Dict[str, types.ModuleType]
+
+    @property
+    def component_type(self) -> ComponentType:
+        """Get the component type."""
+        return self._configuration.component_type
+
+    @property
+    def is_vendor(self) -> bool:
+        """Get whether the component is vendorized or not."""
+        return self._is_vendor
+
+    @property
+    def prefix_import_path(self):
+        """Get the prefix import path for this component."""
+        if self.is_vendor:
+            return "packages.{}.{}.{}".format(
+                self.public_id.author,
+                self.component_type.to_plural(),
+                self.public_id.name,
+            )
+        else:
+            return "packages.{}.{}".format(
+                self.component_type.to_plural(), self.public_id.name
+            )
+
+    @property
+    def component_id(self) -> ComponentId:
+        """Ge the package id."""
+        return self._configuration.component_id
+
+    @property
+    def public_id(self) -> PublicId:
+        """Get the public id."""
+        return self._configuration.public_id
+
+    @property
+    def configuration(self) -> ComponentConfiguration:
+        """Get the component configuration."""
+        return self._configuration
+
+    @property
+    def directory(self) -> Optional[Path]:
+        """Get the directory, or None if the component is not associated with any directory."""
+        return self._directory
+
+    def setup(self, *args, **kwargs):
```

Comment:
What about teardown?

Reply:
we should discuss. `setup` is used below in `load_from_directory`

Reply:
replace with `load`